Better error message for accidental calls to get-accessors

jtbandes opened the pull request:
Fixes #24554 by adding a new Message when `resolveCallExpression` fails to find any call signatures, if the `node.expression` references a `GetAccessor`. Example test case with new message:
```
==== tests/cases/compiler/accessorAccidentalCallDiagnostic.ts (1 errors) ====
    // https://github.com/microsoft/TypeScript/issues/24554
    class Test24554 {
        get property(): number { return 1; }
    }
    function test24554(x: Test24554) {
        return x.property();
                 ~~~~~~~~
!!! error TS2349: This expression is not callable.
!!! error TS2349:   Type 'Number' has no call signatures.
!!! related TS6232 tests/cases/compiler/accessorAccidentalCallDiagnostic.ts:6:12: 'x.property' is a 'get' accessor; did you mean to use it without '()'?
!!! related TS2728 tests/cases/compiler/accessorAccidentalCallDiagnostic.ts:3:9: 'property' is declared here.
    }
```

This is my first attempt to contribute to TS, so I'm eager for any feedback! Specifically here are some questions I'd like someone familiar with the code to answer:
1. Was I right to just choose an available code in diagnosticMessages.json? Are they arranged in some grouping other than by code?
2. Is `resolveCallExpression` a good place to add this? The discussion on #27897 mentioned changing `checkPropertyAccessExpression`, but that seemed hard to me because the PropertyAccessExpression isn't really aware of its parent CallExpression (well, I guess it might be able to access `parent`, but that doesn't seem great—and there are some other diagnostics in this function already).
3. Any caveats to using `getNodeLinks(node.expression).resolvedSymbol`? That was the easiest way I found to reach the function and determine whether it's a GetAccessor, but it also seems to implicitly depend on something else having resolved the symbol first.
4. Is `relatedInformation` the right place to attach such a message? I'm a bit confused about the diagnostic "chain" and how relatedInfo fits in, but this seems to work.
5. I'd be interested in adding a code fix to delete the `()` as well. I started down that path, but it seems like fixes can't be triggered on `relatedInformation`, just on errors (I was able to create a new fix that got triggered on `This_expression_is_not_callable` errors, but it didn't trigger if I register it only for my new diagnostic message.) And since the fix's `getCodeActions(context)` function doesn't receive the diagnostic itself, just the code, it doesn't seem feasible to register it for all `This_expression_is_not_callable` errors and look at their relatedInformation to see if my new `did you mean to use it without '()'?` is attached. One approach I could consider is making it a separate error, but then it wouldn't be attached to the existing "not callable" error. Any advice here?

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/TypeScript?pullRequest=37800) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/TypeScript?pullRequest=37800) before we can accept your contribution.<br/><br/>:x: jtbandes [sign now](https://cla.opensource.microsoft.com/microsoft/TypeScript?pullRequest=37800)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/TypeScript?pullRequest=37800) it.</sub>

DanielRosenwasser commented:
@typescript-bot pack this

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -25771,14 +25771,21 @@ namespace ts {
+                        // Diagnose get accessors incorrectly called as functions
+                        const { resolvedSymbol } = getNodeLinks(node.expression);
+                        if (resolvedSymbol && resolvedSymbol.flags & SymbolFlags.GetAccessor) {
+                            relatedInformation.push(createDiagnosticForNode(node.expression, Diagnostics._0_is_a_get_accessor_did_you_mean_to_use_it_without, getTextOfNode(node.expression)));
```
If you change the error message, you might want to add a diagnostic that points back to the declaration - we already have `Diagnostics._0_is_declared_here`.

jtbandes replied:
I wasn't sure how to get back to the original declaration. Would it be ok to add a special-case check here whether the `node.expression` is a PropertyAccessExpression? (seems a bit roundabout.)

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -4384,6 +4384,10 @@
         "category": "Error",
         "code": 6231
     },
+    "'{0}' is a 'get' accessor; did you mean to use it without '()'?": {
```
```suggestion
    "'{0}' is declared as a 'get' accessor and does not need to be called.": {
```

DanielRosenwasser replied:
Actually, trying this out in an editor, I could see either one being good. It's up to you.

jtbandes replied:
I could see the wording that mentions `()` being better if we could actually provide a code fix that deletes the (). I had some questions about that in my PR description, would you happen to have experience with code fixes and could help me figure out how to implement it cleanly?

sandersn replied:
With or without the codefix, I like the `()` message better since it shows what's wrong instead of naming it in terms that not everyone understands immediately.

sandersn replied:
re codefixes: There are some simple ones that you can basically copy, then you just have to learn how to extract the property access from the call, which should also be easy.

However, I'd rather just take the error alone and follow up with the codefix.

sandersn replied:
Actually, as a replacement for 'this expression is not callable', we should keep some elements of that:
```suggestion
    "This is not callable because it is a 'get' accessor; did you mean to use it without '()'?": {
```

(You already have an error span so I don't think you need to reproduce the expression in the error -- long expressions result in too-long errors.)

sandersn commented on tests/baselines/reference/accessorAccidentalCallDiagnostic.errors.txt:
```diff
@@ -0,0 +1,18 @@
+    function test24554(x: Test24554) {
+        return x.property();
+                 ~~~~~~~~
+!!! error TS2349: This expression is not callable.
```
I would much rather replace this message instead of tacking on related info.

If it's too hard to replace the message, it should be an elaboration, not related info. Elaborations provide additional messages; related info is designed to point to a different location.

sandersn commented:
`src/services/codefixes/fixForgottenThisPropertyAccess.ts` is a good, simple codefix to start from. But it requires a unique error code, which means replacing "This expression is not callable".

sandersn commented:
(Sorry, I didn't read your description until just now -- you asked about every single one of the aspects I commented on.)

jtbandes commented:
Thanks for the feedback!

> But it requires a unique error code, which means replacing "This expression is not callable".

That partially answers my question about the best way to make the code fix. But is it really desirable to replace this error message? I guess I'm worried about the rare cases where the `remove ()` suggestion is wrong, and then the new error message wouldn't really make sense. Should we try to also preserve the "not callable" error (the actual "root cause"), or do you really think it's best to replace it?

jtbandes commented:
I see, I guess starting with "This expression is not callable because…" alleviates my concern. And I suppose we can still keep the "Type {0} has no call signatures" sub-error.

jtbandes commented:
I added a code fix, but to make it easier I modified the parser to save the open/close paren tokens in CallExpression. Is that alright?

jtbandes commented:
Alright, got the code fix working, except that I had to omit the `(`/`)` tokens from forEachChild and change the test harness 😕 : https://github.com/jtbandes/TypeScript/pull/1/files#diff-e4047651bbe0a47423c7db850216a6afR122

jtbandes commented:
Hi @sandersn just a friendly ping on this PR! No rush but I don't want it to languish for too long either :)

sandersn commented on tests/baselines/reference/instancePropertyInClassType.errors.txt:
```diff
@@ -31,8 +29,8 @@ tests/cases/conformance/classes/members/classTypes/instancePropertyInClassType.t
         var r6 = c.y(); // error
                    ~
-!!! error TS2349: This expression is not callable.
-!!! error TS2349:   Type 'Number' has no call signatures.
```
I thought that this elaboration would stay. Why is it gone?

sandersn commented on tests/baselines/reference/instancePropertyInClassType.errors.txt:
```diff
@@ -31,8 +29,8 @@ tests/cases/conformance/classes/members/classTypes/instancePropertyInClassType.t
-!!! error TS2349: This expression is not callable.
-!!! error TS2349:   Type 'Number' has no call signatures.
+!!! message TS6232: This expression is not callable because it is a 'get' accessor. Did you mean to use it without '()'?
+!!! related TS2728 tests/cases/conformance/classes/members/classTypes/instancePropertyInClassType.ts:4:13: 'y' is declared here.
```
I'm not convinced this is valuable. If it is, it should really be on the base "This expression is not callable." error as well. @DanielRosenwasser can you decide on this? I trust your opinion on related spans.

sandersn commented on src/compiler/checker.ts:
```diff
@@ -26127,10 +26127,21 @@ namespace ts {
+                // Diagnose get accessors incorrectly called as functions
+                const { resolvedSymbol } = getNodeLinks(errorTarget);
+                if (resolvedSymbol && resolvedSymbol.flags & SymbolFlags.GetAccessor) {
+                    diagnostic = createDiagnosticForNode(errorTarget, Diagnostics.This_expression_is_not_callable_because_it_is_a_get_accessor_Did_you_mean_to_use_it_without);
```
I think the change should be in invocationErrorDetails, very close to the base "Type_0_has_no_call_signatures", instead of out here in invocationError.

jtbandes replied:
invocationErrorDetails currently doesn't have enough information (access to the node/symbol) to produce the new error. I could rethink its signature, but that seemed more invasive than just putting it here. I'll take another look later if you think it's better that way!

jtbandes commented:
Alright, I've moved the changes into `invocationErrorDetails` and removed the related info. (Considered keeping it but the code flow around chains vs related info is already convoluted and I didn't want to make it worse.) Let me know what you think!

jtbandes commented:
Ping @sandersn 🙌 (Not really in a hurry, I'm sure you are quite busy — just want to avoid falling through the cracks!)

sandersn commented on src/compiler/diagnosticMessages.json:
```diff
@@ -4416,6 +4416,10 @@
         "code": 6233
     },
+    "This expression is not callable because it is a 'get' accessor. Did you mean to use it without '()'?": {
+        "category": "Message",
```
Should be an Error, not a Message.

sandersn commented:
Travis is all green when I click through to it, so I'm going to go ahead and merge.

jtbandes commented:
Great, thanks! I will send out another PR for the code fix in the near future.
